### DISCUSSION TITLE
file_packager should support modularization

### DISCUSSION THREAD

**Discussion by takase1121** [emscripten-core/emscripten#22901] [Ideas]
I'm working on bundling files generated by `file_packager` with my emcc-compiled source and some wrapper code with esbuild. I ran into issues of the loader script generated by `file_packager` defining `Module`, which clobber the global namespace. I am working around this by wrapping `file_packager` with another script that wraps its output with an CommonJS export. It would be great if `file_packager` also supports something similar to `-sMODULARIZE=1`, where it exports a factory function that takes in `Module` and spits out another `Module`.

**Comment by sbc100** [comment]
I assume you are running the file_packager tool standalone and then taking the output and using it via `--pre-js`?  Or are you include it on the page yourself alongside the emscripten output?

I think you are correct that file_packager needs some kind of extra mode here, but it may require user intervention to glue the file_packager code to the module being instantiated.

**Reply by takase1121** [reply]
> I assume you are running the file_packager tool standalone and then taking the output and using it via `--pre-js`? Or are you include it on the page yourself alongside the emscripten output?
> 
> I think you are correct that file_packager needs some kind of extra mode here, but it may require user intervention to glue the file_packager code to the module being instantiated.

I am running file_packager separately, and feeding its output into a file that is included when bundling my own code with esbuild. The idea is that both file_packager output and emscripten generated JS is treated as JavaScript source when bundling. To do this I have MODULARIZE=1 for emscripten, which would export and loaded properly by esbuild.

What I'm suggesting is to add a factory function to the output, similar to emscripten MODULARIZE=1. The factory function will return a Module that can be chained. I currently have a wrapper script that does this, but I think this is worth adding to file_packager itself.

**Reply by takase1121** [reply]
As a side note, I am avoiding using --pre-js because it requires relinking the project, and that takes a while (about 5 to 8 seconds). Previously I've also used closure which increases this number by a bit as well.
